Rolling window endpoints inclusion

carlosdanielcsantos opened the pull request:
- [x] closes #13965 
 - [x] time-based window tests added / passed
 - [x] fixed window tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

carlosdanielcsantos committed:
Adding window slicing endpoint inclusion selection to VariableWindowIndexer

carlosdanielcsantos committed:
Commiting progress

Commit:
Time-based windows working

‘Closed’ feature is working as expected in time-based windows. Sum
tests are passing.

carlosdanielcsantos commented on pandas/core/window.pyx:
```diff
@@ -1141,11 +1181,11 @@ def roll_min(ndarray[numeric] input, int64_t win, int64_t minp,
 
 cdef _roll_min_max(ndarray[numeric] input, int64_t win, int64_t minp,
-                   object index, bint is_max):
+                   object index, str closed, bint is_max):
```
Not sure about the behavior of str here...

carlosdanielcsantos replied:
Not sure about the behavior of str here (cdef)

jreback replied:
same

jreback commented on pandas/core/window.py:
```diff
@@ -101,6 +103,9 @@ def validate(self):
         if self.min_periods is not None and not \
            is_integer(self.min_periods):
             raise ValueError("min_periods must be an integer")
+        if self.closed not in ['right', 'both', 'left', 'neither']:
```
testing this?

jreback commented on pandas/core/window.pyx:
```diff
@@ -161,7 +161,8 @@ cdef class MockFixedWindowIndexer(WindowIndexer):
     def __init__(self, ndarray input, int64_t win, int64_t minp,
-                 object index=None, object floor=None):
+                 object index=None, object floor=None,
+                 bint l_closed=False, bint r_closed=True):
```
if you have a chance, can you add a mini-doc strings to things.

jreback commented on pandas/core/window.pyx:
```diff
@@ -307,18 +322,35 @@ def get_window_indexer(input, win, minp, index, floor=None,
         compat Indexer that allows us to use a standard
         code path with all of the indexers.
```
can you augment the doc-string here (with closed)

jreback commented on pandas/core/window.pyx:
```diff
@@ -307,18 +322,35 @@ def get_window_indexer(input, win, minp, index, floor=None,
+        bint l_closed = False
+        bint r_closed = False
+
+    if closed not in ['right', 'left', 'both', 'neither']:
```
you can just do an assert here (this is an internal routine that has certain guarantees)

jreback commented on pandas/core/window.pyx:
```diff
@@ -327,7 +359,7 @@ def get_window_indexer(input, win, minp, index, floor=None,
 
 def roll_count(ndarray[double_t] input, int64_t win, int64_t minp,
-               object index):
+               object index, closed='right'):
```
just make this required

``object closed``

jreback commented on pandas/core/window.pyx:
```diff
@@ -523,7 +556,7 @@ cdef inline void remove_mean(double val, Py_ssize_t *nobs, double *sum_x,
 
 def roll_mean(ndarray[double_t] input, int64_t win, int64_t minp,
-              object index):
+              object index, str closed):
```
``object closed``

jreback commented on pandas/core/window.pyx:
```diff
@@ -1111,7 +1148,7 @@ cdef inline numeric calc_mm(int64_t minp, Py_ssize_t nobs,
 
 def roll_max(ndarray[numeric] input, int64_t win, int64_t minp,
-             object index):
+             object index, str closed):
```
same for the rest of these: str-> object

jreback commented on pandas/tests/test_window.py:
```diff
@@ -3356,6 +3356,34 @@ def test_min_periods(self):
         tm.assert_frame_equal(result, expected)
 
+    def test_closed(self):
+        df = DataFrame({'A': [1]*5},
```
add this issue number as a comment

jreback commented on pandas/tests/test_window.py:
```diff
@@ -3356,6 +3356,34 @@ def test_min_periods(self):
+        df = DataFrame({'A': [1]*5},
+                        index = [pd.Timestamp('20130101 09:00:01'),
+                                 pd.Timestamp('20130101 09:00:02'),
+                                 pd.Timestamp('20130101 09:00:03'),
```
test passing invalid closed

jreback commented on pandas/tests/test_window.py:
```diff
@@ -3356,6 +3356,34 @@ def test_min_periods(self):
+        expected["A"] = [np.nan, 1.0, 2, 2, 1]
+        result = df.rolling('2s', closed='left').sum()
+        tm.assert_frame_equal(result, expected)
+
```
@chrisaycock how do these tests look? what corners do we need to test here

chrisaycock replied:
What happens if the DataFrame is empty?

carlosdanielcsantos replied:
If the DataFrame is empty, the behavior is exactly the same as in the current implementation. For time-based windows, it throws "ValueError: window must be an integer".  Note that this feature only affects the window indexer's start and end vectors. In fixed windows, empty DataFrames always return empty (the result is constructed as np.empty(N) where N is the length of the input)

jreback replied:
separately I think [4] should just return an empty frame for consistency.

```
In [4]: DataFrame().rolling('1s').sum()
ValueError: window must be an integer

In [5]: DataFrame().rolling(1).sum()
Out[5]: 
Empty DataFrame
Columns: []
Index: []
```

@chrisaycock ?

carlosdanielcsantos commented:
Here is a quick example of what this feature looks like

```
df = pd.DataFrame({'x': [1]*5},
....:                 index = [pd.Timestamp('20130101 09:00:01'),
....:                          pd.Timestamp('20130101 09:00:02'),
....:                          pd.Timestamp('20130101 09:00:03'),
....:                          pd.Timestamp('20130101 09:00:04'),
....:                          pd.Timestamp('20130101 09:00:06')])

df["right"] = df.rolling('2s', closed='right').x.sum()   # current behavior
df["both"] = df.rolling('2s', closed='both').x.sum()
df["left"] = df.rolling('2s', closed='left').x.sum()
df["open"] = df.rolling('2s', closed='none').x.sum()

print(df)
                     x  right  both  left  open
2013-01-01 09:00:01  1    1.0   1.0   NaN   NaN
2013-01-01 09:00:02  1    2.0   2.0   1.0   1.0
2013-01-01 09:00:03  1    2.0   3.0   2.0   1.0
2013-01-01 09:00:04  1    2.0   3.0   2.0   1.0
2013-01-01 09:00:06  1    1.0   2.0   1.0   NaN
```

jreback reviewed (changes requested):
mostly minor points
- add a whatsnew entry (Other Enhancements) in whatsnew (0.20.0), with a link to the computation.rst docs (where you have a new section showing the use of closed; you can use your test examples, e.g. showing all 4 cases).

jreback commented:
@carlosdanielcsantos https://github.com/pandas-dev/pandas/pull/15795#issuecomment-288883027 are basically the docs (with a bit of prose) and nice formatting. Laying out the possibilities (in a sub-section of computation.rst). Note you also might want to have a 'real-world' uses case sentence for some of these.

jreback commented:
you can see pep errors: http://pandas-docs.github.io/pandas-docs-travis/contributing.html#python-pep8

use ``git diff master | flake8 --diff``

carlosdanielcsantos committed:
str closed -> object closed
Adding test of assert for closed parameter
Adding assert for closed parameter in get_window_indexer

carlosdanielcsantos committed:
Adding doc-strings and PEP8 corrections

Also, adding whatsnew entry

carlosdanielcsantos commented:
@jreback How should I add this to computations.rst? A new section inside Window Functions or just add the examples inside the existing Window Functions -> Rolling Windows?

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -268,6 +268,7 @@ To convert a ``SparseDataFrame`` back to sparse SciPy matrix in COO format, you
 Other enhancements
 ^^^^^^^^^^^^^^^^^^
 
+- ``DataFrame.rolling()`` now accepts the parameter ``closed='right'|'left'|'both'|'neither'`` to choose the rolling window endpoint closedness. (:issue:`13965`)
```
is that a word?

carlosdanielcsantos replied:
There is some discussion going on about that online, but its a useful word to reference the property of being closed. But I am open to suggestions, no pun intended :)

jreback replied:
hahah yep that is fine

jreback replied:
maybe ``closed-ness``? (though I don't think *that* is actually a word)

jreback replied:
also add a ``:ref:`` to the new sub-section that you are creating in computation.rst

jreback commented on pandas/core/window.py:
```diff
@@ -374,8 +379,12 @@ class Window(_Window):
     on : string, optional
         For a DataFrame, column on which to calculate
         the rolling window, rather than the index
+    closed : 'right', 'left', 'both', 'neither'
```
add the default setting (right)

jreback replied:
I would do this with a bulleted list

jreback commented on pandas/core/window.pyx:
```diff
@@ -158,10 +158,15 @@ cdef class MockFixedWindowIndexer(WindowIndexer):
     floor: optional
         unit for flooring
+    l_closed: bint
+        left endpoint closedness
```
I would write these out ``l_closed`` -> ``left_closed`` easier to read

jreback commented on pandas/core/window.pyx:
```diff
@@ -229,10 +239,14 @@ cdef class VariableWindowIndexer(WindowIndexer):
     index: ndarray
         index of the input
+    l_closed: bint
+        left endpoint closedness
```
can add a bit of explication here (1 line), just for future readers

jreback commented on pandas/core/window.pyx:
```diff
@@ -261,7 +276,11 @@ cdef class VariableWindowIndexer(WindowIndexer):
 
         start[0] = 0
-        end[0] = 1
+
```
put the comments above the code

```
if r_closed:
    # .....
    end[0] = 1
else:
    # ....
    end[0]  0
```

jreback commented on pandas/core/window.pyx:
```diff
@@ -271,6 +290,9 @@ cdef class VariableWindowIndexer(WindowIndexer):
                 end_bound = index[i]
                 start_bound = index[i] - win
 
+                if l_closed:  # left endpoint is closed
```
same

jreback replied:
of can do (I use this if only an if statement)
```
# left endpoint ....
if left_closed:
    ....
```

jreback commented on pandas/core/window.pyx:
```diff
@@ -286,9 +308,13 @@ cdef class VariableWindowIndexer(WindowIndexer):
                 else:
                     end[i] = end[i - 1]
 
+                # right endpoint is open
```
yep excatly like this one

jreback commented on pandas/core/window.pyx:
```diff
@@ -299,6 +325,8 @@ def get_window_indexer(input, win, minp, index, floor=None,
     minp: integer, minimum periods
     index: 1d ndarray, optional
         index to the input array
+    closed: 'right', 'left', 'both', 'neither'
```
```
closed: string, default 'right'
        {'right',......}
```

jreback commented:
@carlosdanielcsantos lgtm. some more doc-string nitpicks.

jreback commented:
ping when all green.

carlosdanielcsantos commented:
@jreback I was going to start writing tests for the FixedWindowIndexer but there are no explicit rolling computing test going on in TestRolling (they are done in TestMoments, I think). Where should I add these new tests?

jreback commented:
> @jreback How should I add this to computations.rst? A new section inside Window Functions or just add the examples inside the existing Window Functions -> Rolling Windows?

exactly, a new sub-section that pretty much shows the example with all 4 closed. maybe right before Centering Windows?

jreback commented:
> @jreback I was going to start writing tests for the FixedWindowIndexer but there are no explicit rolling computing test going on in TestRolling (they are done in TestMoments, I think). Where should I add these new tests?

We don't have explicit tests for the cython routines, these are covered by the integration tests (top-level routines). Though certainly could add them if you want / can. 

You could add a new class in test_windows that explicity exercises these.

carlosdanielcsantos commented:
@jreback @chrisaycock 
Thinking deeper about the fixed window case, it seems that, intuitively, the default behavior right now better resembles the closed='both' case. Notice that when you use rolling(3).apply(f), the result for index i will be the result of f applied to `[i-2, i-1, i]`, not `]i-2, i-1, i]` (ISO 31-11 math set notation). One could argue that this is equivalent to having closed='right' with f applied to `]i-3, i-2, i-1, i]`. Indeed, but this probably adulterates the meaning of the window size.

jreback commented:
> @jreback @chrisaycock
Thinking deeper about the fixed window case, it seems that, intuitively, the default behavior right now better resembles the closed='both' case. Notice that when you use rolling(3).apply(f), the result for index i will be the result of f applied to [i-2, i-1, i], not ]i-2, i-1, i] (ISO 31-11 math set notation). One could argue that this is equivalent to having closed='right' with f applied to ]i-3, i-2, i-1, i]. Indeed, but this probably adulterates the meaning of the window size.

@carlosdanielcsantos can you show a comparison of these 2, to show more what you mean?

jreback commented:
> Nonetheless, I think that the possibility of ignoring the current index in computations is valuable (that was my initial problem, after all). Some solutions for this are:

This a bit orthogonal though. We *always* want to actually use the index. Since using a fixed window size, it *looks* like we are ignoring the index :>.

So a possiblity is to actually check the index whether its equiv to a RangeIndex. e.g.

```
In [5]: df1 = pd.DataFrame([1] * 5)

In [6]: df1.rolling(3, min_periods=1).sum()
Out[6]: 
     0
0  1.0
1  2.0
2  3.0
3  3.0
4  3.0

In [8]: df2 = pd.DataFrame([1] * 5, index=[0, 1, 3, 4, 5])

In [9]: df2.rolling(3, min_periods=1).sum()
Out[9]: 
     0
0  1.0
1  2.0
3  3.0
4  3.0
5  3.0
```

```
In [11]: (np.diff(df1.index)==1).all()
Out[11]: True

In [12]: (np.diff(df2.index)==1).all()
Out[12]: False

```

Let's deal with this in a separate issue. 

A way to disambiguate on this ATM is to allow ``closed=None`` to be the default parameter. Then it cannot be anything but that for a FixedWindow. For Variable, if we see ``closed=None`` -> ``closed='right'`` IOW the default is the same. And then allow the rest of the closed intervals.

I think *we* should fix this for Fixed intervals, but I agree this is a problem because we essentially have ``closed='both'`` as the default now (its implicit though) (as you stated above).

jreback commented:
@carlosdanielcsantos if you can update would be great (docs and such), ignoring the issue about fixed windows for the moment.

carlosdanielcsantos commented:
I'll do that in the next hours. Should I just leave the fixed-window as it is?

jreback commented:
I think make the ``closed=None`` the default, and document the possibilities (and default) for the variable sized window (and validate left,right,both,neither). For the fixed window document as well (you can say its closed='both', but ATM only allow None). and then open a new issue where we can deal with this.

We have to fix this, but this PR merely opens up the possibilities of change here. So I think ok to merge this.

carlosdanielcsantos committed:
Changing l_closed and r_closed variable names

carlosdanielcsantos committed:
Commiting progress on default=None. Still not tested
Adding computation.rst section (still not written)

Commit:
Almost there

Not allowing closed parameter for fixed windows
Passing tests again
Updating docs

jreback commented:
@carlosdanielcsantos can you update

Commit:
Correcting bug in window validation

Commit:
Fixing test of assertion of closed parameter in fixed windows
Style corrections

Commit:
Fixing style

carlosdanielcsantos commented:
@jreback made some corrections but circleci is not passing in test_append_index - pandas.tests.test_multilevel.TestMultiLevel, but passing locally. Any thoughts?

jreback commented:
try pushing again (make a modification)
looks like a fluke

Commit:
Innocuous change to rerun tests

carlosdanielcsantos commented:
@jreback Still not passing. Could I have broken this with the changes in window?

jreback commented:
@carlosdanielcsantos I rebased you against current master. This should pass cleanly. If you'd make those other changes and ping.

carlosdanielcsantos commented:
@jreback what changes? I believe I updated everything that you requested. Can you take a look at AppVeyor build that failed?

jreback commented on doc/source/computation.rst:
```diff
@@ -459,6 +459,43 @@ default of the index) in a DataFrame.
+.. _stats.rolling_window.endpoints:
+
+Rolling window endpoint inclusion
+~~~~~~~~~~~~~~~~~~
```
make this Rolling Window Endpoints

jreback commented on doc/source/computation.rst:
```diff
@@ -459,6 +459,43 @@ default of the index) in a DataFrame.
+The inclusion of the interval endpoints in rolling window calculations can be specified with the ``closed``
+parameter:
+
+  - ``right`` : close right endpoint (default for time-based windows)
```
can you make this a csv-table (3 columns: ``closed``, ``description``, ``default``) or something like that

jreback commented on doc/source/computation.rst:
```diff
@@ -459,6 +459,43 @@ default of the index) in a DataFrame.
+
+   df
+
+Currently, this feature is only implemented for time-based windows.
```
expand this a touch to say that *only* ``closed='both'`` is accepted for fixed windows.

carlosdanielcsantos replied:
I am only accepting closed=None for fixed windows, as previously requested by you

> For the fixed window document as well (you can say its closed='both', but ATM only allow None)

jreback replied:
that's what i mean just make sure the docs are clear on that ; e.g. None -> 'both' for fixed (and assert that as well )

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -319,6 +319,7 @@ To convert a ``SparseDataFrame`` back to sparse SciPy matrix in COO format, you
 Other Enhancements
 ^^^^^^^^^^^^^^^^^^
 
+- ``DataFrame.rolling()`` now accepts the parameter ``closed='right'|'left'|'both'|'neither'`` to choose the rolling window endpoint closedness. See :ref:`documentation <stats.rolling_window.endpoints>` (:issue:`13965`)
```
See ``:ref:`the documentation ....``

jreback commented on pandas/core/window.py:
```diff
@@ -101,6 +103,10 @@ def validate(self):
             raise ValueError("min_periods must be an integer")
+        if self.closed is not None and self.closed not in \
+           ['right', 'both', 'left', 'neither']:
+            raise ValueError("closed must be 'right', 'left', 'both' or "
```
is there validation for a fixed window when closed is NOT both?

carlosdanielcsantos replied:
Again, there is a validation for when closed is not None in Rolling.validate()

jreback reviewed (approved):
minor comments.

Commit:
Updating docs

carlosdanielcsantos commented on pandas/tests/test_window.py:
```diff
@@ -431,6 +431,12 @@ def test_numpy_compat(self):
             tm.assertRaisesRegexp(UnsupportedFunctionCall, msg,
                                   getattr(r, func), dtype=np.float64)
 
+    def test_closed(self):
```
@jreback here is the piece that tests that closed must be None for fixed-window. Well, actually it only tests that it should not be 'neither', but I do not think replication is necessary.

carlosdanielcsantos commented on pandas/core/window.py:
```diff
@@ -1044,6 +1058,10 @@ def validate(self):
         elif self.window < 0:
             raise ValueError("window must be non-negative")
 
+        if not self.is_datetimelike and self.closed is not None:
```
@jreback here we enforce that `closed` must be `None` for windows that are not datetimelike

jreback commented:
thanks @carlosdanielcsantos nice PR. as you can see, the code changes are the 'easy' part, its the docs & testing that are the details.!

keem em coming!
